tests/std/tests/P0896R4_views_join/test.cpp

```diff
@@ -0,0 +1,505 @@
+
+            static_assert(same_as<decltype(views::join(move(as_const(rng)))), R>);
+            static_assert(noexcept(views::join(move(as_const(rng)))) == is_noexcept);
+
+            static_assert(same_as<decltype(move(as_const(rng)) | views::join), R>);
+            static_assert(noexcept(move(as_const(rng)) | views::join) == is_noexcept);
+        } else if constexpr (!is_view && borrowed_range<const remove_cvref_t<Outer>>) {
+            using S                    = decltype(ranges::subrange{as_const(rng)});
+            using RS                   = join_view<S>;
+            constexpr bool is_noexcept = noexcept(S{as_const(rng)});
+
+            static_assert(same_as<decltype(views::join(move(as_const(rng)))), RS>);
+            static_assert(noexcept(views::join(move(as_const(rng)))) == is_noexcept);
+
+            static_assert(same_as<decltype(move(as_const(rng)) | views::join), RS>);
+            static_assert(noexcept(move(as_const(rng)) | views::join) == is_noexcept);
+        }
+
+        // Validate deduction guide
+        same_as<R> auto r = join_view{forward<Outer>(rng)};
+        assert(ranges::equal(r, expected));
+        const bool is_empty = ranges::empty(expected);
+
+        // Validate lack of size
+        static_assert(!CanSize<R>);
+
+        // Validate view_interface::empty and operator bool
+        static_assert(CanEmpty<R> == forward_range<R>);
+        static_assert(CanMemberEmpty<R> == CanEmpty<R>);
+        if constexpr (CanMemberEmpty<R>) {
+            assert(r.empty() == is_empty);
+            assert(static_cast<bool>(r) == !is_empty);
+
+            static_assert(CanEmpty<const R> == forward_range<const R>);
+            static_assert(CanMemberEmpty<const R> == CanEmpty<const R>);
+            if constexpr (CanMemberEmpty<const R>) {
+                assert(as_const(r).empty() == is_empty);
+                assert(static_cast<bool>(as_const(r)) == !is_empty);
+            }
+        }
+
+        // Validate join_view::begin
+        static_assert(CanMemberBegin<R>);
+        static_assert(CanMemberBegin<const R> == (input_range<const V> && is_reference_v<range_reference_t<const V>>) );
+        if constexpr (forward_range<R>) {
+            const iterator_t<R> i = r.begin();
+            if (!is_empty) {
+                assert(*i == *begin(expected));
+            }
+
+            if constexpr (copyable<V>) {
+                auto r2                              = r;
+                const same_as<iterator_t<R>> auto i2 = r2.begin();
+                if (!is_empty) {
+                    assert(*i2 == *i);
+                }
+            }
+
+            if constexpr (CanMemberBegin<const R> && !CanBegin<const R>) {
+                static_assert(input_iterator<decltype(declval<const R>().begin())>);
+                static_assert(sentinel_for<decltype(declval<const R>().end()), decltype(declval<const R>().begin())>);
+            }
+
+            if constexpr (CanBegin<const R> && forward_range<R>) {

```

Comment:
I think there was a `}` here at some point. Otherwise the `&& forward_range<R>` does nto make sense. I think I would somply remove it and keep the rest as is?

Reply:
Oh yeah: this is all inside `if constexpr (forward_range<R>)` from line 159. Striking.